scripts/annotate_sam_with_bed_features.py

```diff
@@ -0,0 +1,328 @@
+#!/usr/bin/env python
+
+# pylint: disable=line-too-long
+
+"""Annotate SAM alignments with their intersecting feature(s).
+
+Add a custom tag ("YW") to each alignment in the SAM file if an intersecting
+feature is found in the BED file. The BED file must be the result of a
+bedtools intersect call where `-a` is a GFF3 file and `-b` a BAM file. If
+either the BED or the SAM file is empty, only the SAM file header is returned.
+
+Each matching feature is used to build a tag with the following format:
+
+    FEATURE_ID|5'-shift|3'-shift|CIGAR|MD|READ_SEQ.
+
+Where:
+    - FEATURE_ID: Extracted from the specified attribute in the BED file
+    (default: "name").
+    - 5'-shift: Difference between the (possibly adjusted) feature start and
+    the alignment start.
+    - 3'-shift: Difference between the alignment end and the (possibly
+    adjusted) feature end.
+    - CIGAR and MD: The alignment's CIGAR string and MD tag respectively.
+    - READ_SEQ: The read sequence from the alignment.
+
+Optional adjustments:
+--extension: Adjust the feature's start and end coordinates by the given value
+    (start is increased and end decreased).
+--shift: Requires both the 5' and 3' shift values to be within +/- this value
+    to include the feature tag in the alignment.
+
+If an alignment has multiple intersecting features, the tag values are
+concatenated using a semicolon as the separator. If no valid intersecting
+feature is found, the alignment is skipped.
+
+Examples
+--------
+Example 1: No tag added due to shift mismatch
+    BED record:
+        19	.	miRNA	44377	44398	.	+	.	ID=MIMAT0002849;Alias=MIMAT0002849;Name=hsa-miR-524-5p;Derives_from=MI0003160	19	44376	44398	13-1_1	1	+	22
+    SAM record:
+        13-1_1	0	19	44377	1	11M3I11M	*	0	0	CTACAAAGGGAGGTAGCACTTTCTC	*	HI:i:0	MD:Z:22	NH:i:1	NM:i:3	RG:Z:A1	YZ:Z:0
+    Command:
+        annotate_sam_with_bed_features.py -b BED -s SAM
+    Outcome:
+        No tag is added because the alignment's 3' end shift exceeds the
+        allowed range.
+
+Example 2: Tag added without coordinate adjustments
+    BED record:
+        19	.	miRNA	5338	5359	.	+	.	ID=MIMAT0005795;Alias=MIMAT0005795;Name=hsa-miR-1323;Derives_from=MI0003786	19	5337	5358	48-1_1	255	+	21
+    SAM record:
+        48-1_1	0	19	5338	255	21M	*	0	0	TCAAAACTGAGGGGCATTTTC	*	MD:Z:21	NH:i:1	NM:i:0
+    Command:
+        annotate_sam_with_bed_features.py -b BED -s SAM
+    Outcome:
+        The tag "hsa-miR-1323|0|0|21M|21|TCAAAACTGAGGGGCATTTTC" is appended
+        because the alignment perfectly matches the feature coordinates.
+
```

Comment:
I think examples would be useful that (1) use only `--shift` and (2) use only `--extension`. It really doesn't help clarity that the available examples use either none or both.

Also, I think the examples would benefit from:
- Stating the use case in words (to clarify the use of `--shift`, `--extensions`) etc.
- Follow the use case with the command.
- _Then_ show the BED and SAM records.
- Then add an actual alignment to visualize what is happening here.
- Add the actual output record.
- Finally add the verbal description/reasoning for the output.